### PR TITLE
Support recursive types

### PR DESCRIPTION
**Pull request by bkchr**

### CODE REVIEW

File: derive/src/trait_bounds.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+				Fields::Unit => { Vec::new() },
+			}
+		}).collect(),
+
+		Data::Union(_) => return Err(Error::new(Span::call_site(), "Union types are not supported.")),
+	};
+
+	Ok(types)
+}
```
Comment (0x7CFE) [inline derive/src/trait_bounds.rs]: newline missing

File: tests/mod.rs [hunk 2]
```diff
@@ -398,3 +395,27 @@ fn generic_trait() {
...
+fn recursive_variant_2_encode_works() {
+	#[derive(Debug, PartialEq, Encode, Decode, Default)]
+	struct Recursive<A, B, N> {
+		data: N,
+		other: Vec<Struct<A, B, Recursive<A, B, N>>>,
+	}
+
+	let val: Recursive<u32, i32, u32> = Recursive::default();
+	val.encode();
+}
```
Comment (0x7CFE) [inline tests/mod.rs]: newline missing

File: derive/src/lib.rs [hunk 3]
```diff
@@ -139,73 +148,4 @@ pub fn decode_derive(input: TokenStream) -> TokenStream {
...
-				Fields::Unit => { Vec::new() },
-			}
-		}).collect(),
-
-		Data::Union(_) => return Err(Error::new(Span::call_site(), "Union types are not supported.")),
-	};
-
-	Ok(types)
-}
+}
```
Comment (0x7CFE) [inline derive/src/lib.rs]: newline missing

**Comment by bkchr** [comment]
In the case of an alias we can not do anything... The user will need to switch to a custom implementation, but I think that this is a really really rare case.

**Pull request merged**